grisp/grisp_base/common/build/nifs/grisp_nspi_nif.c

```diff
@@ -0,0 +1,137 @@
+                                            .spi_regs = SPI0};
+
+static void spi_data_dtor(ErlNifEnv *env, void *obj) {
+    close(((grisp_spi_data *)obj)->fd);
+}
+
+int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info) {
+    atom_ok = enif_make_atom(env, "ok");
+    atom_error = enif_make_atom(env, "error");
+    spi_data = enif_open_resource_type(env, NULL, "spi_data",
+                                       (ErlNifResourceDtor *)spi_data_dtor,
+                                       ERL_NIF_RT_CREATE, NULL);
+
+    assert(spi_data != NULL);
+
+    int rv;
+
+    /* bus registration */
+    rv = spi_bus_register_atsam(ATSAM_SPI_0_BUS_PATH, &spi_config);
+
+    if (rv != 0) perror("SPI bus registration failed");
+    assert(rv == 0);
+
+    return 0;
+}
+
+static ERL_NIF_TERM error_tuple(ErlNifEnv *env, char *reason) {
+    return enif_make_tuple2(env, atom_error, enif_make_atom(env, reason));
+}
+
+static ERL_NIF_TERM open_spi(ErlNifEnv *env, int argc,
+                             const ERL_NIF_TERM argv[]) {
+    ERL_NIF_TERM ret;
+    uint32_t speed = 100000;
+
+    printf(stderr, "Tying to alloc resource");
+    grisp_spi_data *data =
+        enif_alloc_resource(spi_data, sizeof(grisp_spi_data));
+
+    data->fd = open(ATSAM_SPI_0_BUS_PATH, O_RDWR);
+
+    if (data->fd == -1) return error_tuple(env, "open_spi");
+
+    if (ioctl(data->fd, SPI_IOC_WR_MAX_SPEED_HZ, &speed) != 0)
+        return error_tuple(env, "ioctl");
+
+    ret = enif_make_resource(env, data);
+    enif_release_resource(data);
+
+    return ret;
+}
+
+static ERL_NIF_TERM command(ErlNifEnv *env, int argc,
+                            const ERL_NIF_TERM argv[]) {
+    int rv;
+    int cs;
+    int mode;
+    grisp_spi_data *data;
+    ErlNifBinary buf;
+    ErlNifBinary res;
+
+    struct spi_ioc_transfer msg;
+
+    if (!enif_get_resource(env, argv[0], spi_data, (void **)&data))
+        return error_tuple(env, "badarg_ref");
+    if (!enif_get_int(env, argv[1], &cs)) return error_tuple(env, "badarg_cs");
+
+    if (!enif_get_int(env, argv[2], &mode))
+        return error_tuple(env, "badarg_mode");
+
+    if (!enif_inspect_binary(env, argv[3], &buf))
+        return error_tuple(env, "badarg_bin");
+
+    if (!enif_alloc_binary(buf.size, &res))
+        return error_tuple(env, "alloc_bin");
+
+    msg.speed_hz = 100000;
+    msg.delay_usecs = 1;
+    msg.bits_per_word = 8;
+    msg.cs_change = 1;
+    msg.rx_nbits = 0;
+    msg.tx_nbits = 0;
+
+    msg.cs = cs;
+    msg.mode = ((mode & CPOL_HIGH) ? SPI_CPOL : 0) |
+               ((mode & CPHA_TRAILING) ? SPI_CPHA : 0);
+
+    msg.tx_buf = buf.data;
+    msg.rx_buf = res.data;
+    msg.len = buf.size;
+
+    rv = ioctl(data->fd, SPI_IOC_MESSAGE(1), &msg);
+    if (rv != 0) return error_tuple(env, "ioctl");
+
+    return enif_make_binary(env, &res);
+}
+
+static ErlNifFunc nif_funcs[] = {{"open", 0, open_spi},
+                                 {"command_nif", 4, command}};
```

Comment:
We should probably figure out a good naming scheme here to make grepping the source code easier. Currently we have a mix of `function`, `function_nif` and `function_xxx` (e.g. to not clash with `open`). Would be best if all Erlang _and_ C functions are called the same I think.

Ending with `_nif` would be the obvious choice.